File: centrifuge-RemoveN.cpp
#include <stdio.h>
#include <stdlib.h>

#define BUFFER_SIZE 1000000

char usage[] = "a.out xxx.fa" ;

char buffer[ BUFFER_SIZE + 1 ] ;

int main( int argc, char *argv[] )
{
	FILE *fp = fopen( argv[1], "r" ) ;
	int i, j, k ;
	int bs, l ;
	int state ; // 0-header state, 1-sequence
	
	if ( argc < 2 )
	{
		printf( "%s\n", usage ) ;
		exit( 1 ) ;
	}
	bs = -1 ;
	i = 0 ;
	l = 0 ;
	state = 0 ;
	while ( 1 )	
	{
		if ( i >= bs )
		{
			if ( ( bs = fread( buffer, sizeof( char ), BUFFER_SIZE, fp ) ) == 0 )
				break ;

			i = 0 ;
		}
		if ( buffer[i] == '>' )
		{
			if ( l > 0 )
				printf( "\n" ) ;

			state = 0 ;
			l = 0 ;
		}
		else if ( buffer[i] == '\n' && state == 0 )
		{
			state = 1 ;
			l = -1 ;
		}
		else if ( state == 1 && ( buffer[i] == 'N' || buffer[i] == 'n' || buffer[i] == '\n' ) )
		{
			//printf( "hi\n" ) ;
			++i ;
			continue ;
		}
		
		printf( "%c", buffer[i] ) ;
		++l ;
		if ( l == 80 && state == 1 )
		{
			printf( "\n" ) ;
			l = 0 ;
		}

		++i ;
	}
	return 0 ;
}

File: centrifuge-RemoveN.pl
#!/usr/bin/env perl

use strict ;
use warnings ;

die "usage: a.pl xxx.fa > output.fa" if ( @ARGV == 0 ) ;

my $LINE_WIDTH = 80 ;
my $BUFFER_SIZE = 100000 ;
open FP1, $ARGV[0] ;
my $buffer = "" ;

while ( <FP1> )
{
	if ( /^>/ )
	{
		my $line = $_ ;
		if ( $buffer ne "" )
		{
			my $len = length( $buffer ) ;
			for ( my $i = 0 ; $i < $len ; $i += $LINE_WIDTH )
			{
				print substr( $buffer, $i, $LINE_WIDTH ), "\n" ;
			}
			$buffer = "" ;
		}
		print $line ;
	}
	else
	{
		chomp ;
		tr/nN//d ;
		#my $line = uc( $_ ) ;
		my $line = $_ ;
		$buffer .= $line ;

		if ( length( $buffer ) > $BUFFER_SIZE )
		{
			my $len = length( $buffer ) ;
			my $i = 0 ;
			for ( $i = 0 ; $i + $LINE_WIDTH - 1 < $len ; $i += $LINE_WIDTH )
			{
				print substr( $buffer, $i, $LINE_WIDTH ), "\n" ;
			}
			substr( $buffer, 0, $i, "" ) ;
		}
	}
}
if ( $buffer ne "" )
{
	my $len = length( $buffer ) ;
	for ( my $i = 0 ; $i < $len ; $i += $LINE_WIDTH )
	{
		print substr( $buffer, $i, $LINE_WIDTH ), "\n" ;
	}
	$buffer = "" ;
}
